go/controller/internal/autogen/autogen_api_translator.go

```diff
@@ -87,21 +88,131 @@ func (a *apiTranslator) TranslateToolServer(ctx context.Context, toolServer *v1a
 	}, nil
 }
 
-func translateToolServerConfig(config v1alpha1.ToolServerConfig) (string, *api.ToolServerConfig, error) {
+// resolveValueSource resolves the value from a ConfigMap or Secret
+func (a *apiTranslator) resolveValueSource(ctx context.Context, source *v1alpha1.ValueSource, defaultNamespace string) (string, error) {
+	if source == nil {
+		return "", fmt.Errorf("source cannot be nil")
+	}
+
+	namespace := defaultNamespace
+	if source.Namespace != "" {
+		namespace = source.Namespace
+	}
+
+	name := types.NamespacedName{Name: source.Name, Namespace: namespace}
+
+	return a.fetchValueFromSource(ctx, name, source.Key)
+}
+
+// fetchValueFromSource fetches a value from either a ConfigMap or Secret
+func (a *apiTranslator) fetchValueFromSource(ctx context.Context, name types.NamespacedName, key string) (string, error) {
+	// Try to fetch from ConfigMap first
+	configMap := &corev1.ConfigMap{}
+	err := a.kube.Get(ctx, name, configMap)
+	if err == nil {
+		value, exists := configMap.Data[key]
+		if !exists {
+			return "", fmt.Errorf("key %s not found in ConfigMap %s/%s", key, name.Namespace, name.Name)
+		}
+		return value, nil
+	}
+
+	// If not found in ConfigMap, try Secret
+	secret := &corev1.Secret{}
+	err = a.kube.Get(ctx, name, secret)
+	if err == nil {
+		value, exists := secret.Data[key]
+		if !exists {
+			return "", fmt.Errorf("key %s not found in Secret %s/%s", key, name.Namespace, name.Name)
+		}
+		return string(value), nil
+	}
+
+	return "", fmt.Errorf("failed to find ConfigMap or Secret %s/%s: %v", name.Namespace, name.Name, err)
+}
+
+func (a *apiTranslator) translateToolServerConfig(ctx context.Context, config v1alpha1.ToolServerConfig, namespace string) (string, *api.ToolServerConfig, error) {
 	switch {
 	case config.Stdio != nil:
+		// Start with the base environment variables
+		env := make(map[string]string)
+
+		// Add direct environment variables first
+		if config.Stdio.Env != nil {
+			for k, v := range config.Stdio.Env {
+				env[k] = v
+			}
+		}
+
+		// Add environment variables from EnvFrom
+		if len(config.Stdio.EnvFrom) > 0 {
+			for _, envVar := range config.Stdio.EnvFrom {
+				if envVar.ValueFrom != nil {
+					var value string
+					var err error
+
+					if envVar.ValueFrom.ConfigMapKeyRef != nil {
+						value, err = a.resolveValueSource(ctx, envVar.ValueFrom.ConfigMapKeyRef, namespace)
```

Comment:
since you're explicitly checking ConfigMap and SecretKeyRef here, perhaps having a a.getConfigMapValue and getSecretValue (which I think we might already have), might work better.